### PR TITLE
Master server list

### PR DESCRIPTION
**Pull request by zsilencer**
Allows people to click Start Server and specify name, password, port, max users, and whether to advertise to master server url every 60 seconds with the number of players online etc.  You can find the master server php script on https://github.com/OpenRCT2/master-server, it requires sqlite3 and sockets, and specify url in OPENRCT2_MASTER_URL to set it to a default for a build server.  Or player can set in config.  It will try to connect to openrct2 to verify that the port is open and get the server info that way.  It would be easy to extend to allow multiple interfaces listening etc since it uses json.

The multiplayer server list is updated when it is opened or when Fetch Servers is clicked on, and there is a dropdown to join or add/remove from favorites.  There are also some bug fixes.

### CODE REVIEW

**Comment by pizza2004** [comment]
Oh thank goodness this is exactly what I was hoping for in the multiplayer server related things (namely setting all the stuff when you start a server, because knowing what port you were starting it on wasn't intuitive before) but didn't want to say for fear of being told it wasn't important right now!

File: src/interface/chat.c [hunk 1]
```diff
@@ -102,7 +103,7 @@ void chat_history_add(const char *src)
 {
 	int index = _chatHistoryIndex % CHAT_HISTORY_SIZE;
 	memset(_chatHistory[index], 0, CHAT_INPUT_SIZE);
-	memcpy(_chatHistory[index], src, min(strlen(src), CHAT_INPUT_SIZE));
+	memcpy(_chatHistory[index], src, min(strlen(src), CHAT_INPUT_SIZE - 1));
```
Comment (janisozaur) [inline src/interface/chat.c]: `safe_strncpy` instead?

File: src/network/network.cpp [hunk 2]
```diff
@@ -1182,15 +1312,21 @@ int Network::Client_Handle_PINGLIST(NetworkConnection& connection, NetworkPacket
 
 int Network::Client_Handle_SETDISCONNECTMSG(NetworkConnection& connection, NetworkPacket& packet)
 {
-	static char msg[256] = {0};
+	static std::string msg;
 	const char* disconnectmsg = packet.ReadString();
 	if (disconnectmsg) {
-		safe_strncpy(msg, disconnectmsg, sizeof(msg));
-		connection.last_disconnect_reason = msg;
+		msg = disconnectmsg;
```
Comment (janisozaur) [inline src/network/network.cpp]: `disconnectmsg` can be not null-terminated

Comment (zsilencer) [inline src/network/network.cpp]: ?
ReadString always returns a null terminated string

**Comment by IntelOrca** [comment]
Looks good and solid. I will try this out tonight.
One suggestion is to make it default to **OPENRCT2_MASTER_URL** if the url in the config is empty.

**Comment by JarnoVgr** [comment]
Sorta related: Made the openrct2.com build server compatible with this PR for when it gets merged.

( https://openrct2.com/master-server/MasterServer.php )

**Comment by runelaenen** [comment]
openrct.net is also available: https://openrct.net/MS
Public list on https://openrct.net/servers.php

@janisozaur I did edited the php code you provided. I removed all calls with 'key' since it's never mentioned in this pull's code, and because it trew errors. (my code: https://gist.github.com/cab6085043f18ab89c7b it has been edited to work with my database (MySQL) and configuration so it's not very portable to another server, but I think you'll understand the changes.)

Can I also recommend you to also call the server when another player joins? Because right now, the number of players is not correct.

**Comment by IntelOrca** [comment]
We shouldn't need more than one master server, therefore the official one is located at https://servers.openrct2.website. It is fine to still create a third party HTML view of the server list by requesting that URL with the header ```Accept: application/json```. You can also still use a third party master server for OpenRCT2 using the configuration option: ```master_server_url``` which will override the default one.

**Comment by pizza2004** [comment]
@IntelOrca I... but... what about the commits to improve the start server dialog?  I don't understand why this was closed, as what you said doesn't seem to have anything to do with the pull request.

**Comment by JarnoVgr** [comment]
Looks like it has been merged.

**Comment by duncanspumpkin** [comment]
@pizza2004 it was merged just not through this PR.

**Comment by pizza2004** [comment]
Ah, okay.  That makes sense!

**Pull request closed without merging**